src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParameterPart.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    [DebuggerDisplay("{DebuggerToString()}")]
+    public class RoutePatternParameterPart : RoutePatternPart
+    {
+        internal RoutePatternParameterPart(
+            string rawText,
+            string name,
+            object defaultValue,
+            RoutePatternParameterKind parameterKind,
+            RoutePatternConstraintReference[] constraints)
+            : base(RoutePatternPartKind.Parameter, rawText)
+        {
+            // See #475 - this code should have some asserts, but it can't because of the design of RouteParameterParser.
+
+            Name = name;
+            DefaultValue = defaultValue;
+            ParameterKind = parameterKind;
+            Constraints = constraints;
+        }
+
+        public IReadOnlyList<RoutePatternConstraintReference> Constraints { get; }
+
+        public object DefaultValue { get; }
+
+        public bool IsCatchAll => ParameterKind == RoutePatternParameterKind.CatchAll;
+
+        public bool IsOptional => ParameterKind == RoutePatternParameterKind.Optional;
+
+        public RoutePatternParameterKind ParameterKind { get; }
+
+        public string Name { get; }
+
+        internal override string DebuggerToString()
+        {
+            return RawText ?? "{" + (IsCatchAll ? "*" : string.Empty) + Name + (IsOptional ? "?" : string.Empty) + "}";
```

Comment:
Since you're here, you could copy and paste this logic: https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/Internal/RouteTemplateWriter.cs

Reply:
Sure 👍